src/main/java/com/squareup/kotlinpoet/TypeSpec.kt

```diff
@@ -441,6 +444,14 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
     fun addSuperinterface(superinterface: KClass<*>)
         = addSuperinterface(superinterface.asTypeName())
 
+    fun addSuperinterface(superinterface: ParameterSpec) = apply {
```

Comment:
But you'll still have to pass the actual `ParameterSpec` with the primary constructor definition. To print the `by` part, we only need to know the name of that `ParameterSpec`.